### ISSUE TITLE
BatchedEffects

### ISSUE THREAD

**Issue by aaronjensen** [matthieu-beteille/redux-data-fx#1]
Hi there,

I like the look of this. Referring to effects as keys rather than functions gets around some of the problems with using functions ([like having to resist the urge to curry them](https://github.com/redux-loop/redux-loop#easily-test-reducer-results)). 

I did want to make one comment on your api, however, and that's the use of objects to represent a list of effects. It seems that you also allow for an array of objects, but your types at least only allow for that as a result of `combineReducers`.  If you're relying on object keys to indicate which effect to use, you are limiting people to only one effect per effect type. That may be fine in most cases, but it seems an unnecessary limitation. 

Perhaps if effects were specified as an array of tuples: `[['fetch', {...}], ...]` or similar to FSAs: `[{effect: 'fetch', params: {...}}]`. The latter is more verbose, but I could imagine people writing "Effect Creators" (similar to Action Creators) that are pure functions to build effects if desired.

Cheers!

**Comment by mt-bt** [comment]
Hi,

Thanks for the feedback 🙇 

Hmm, I might have gotten the types wrong then (I'll double check that), but you should definitely be able to return a batch of effects from your reducers, in fact there's an example of that in the tests:

https://github.com/matthieu-beteille/redux-data-fx/blob/b9e3b14f72445dcb48f4b646ff7f7cb059efafde/test/redux-data-fx.test.ts#L110-L136

This way you can run more than one effect per effect type.

If you're not a big fan of this syntax, you also have the freedom to implement your effect handler so it can either accept an object or an array of objects:

```javascript
store.registerFX('fetch', (params, getState, dispatch) => {
 if (Array.isArray(params)) {
   params.forEach(params => runEffect(params));
 } else {
   runEffect(params);
 }
})
```

Does that answer your question? Would you see any other benefits of using tuples instead of an object?

**Comment by aaronjensen** [comment]
@matthieu-beteille yes, if it's just the types that are wrong (as it seems to be) then I can already use the library as I'd want to. 

I think from an API design standpoint I'd still prefer to see one way to emit multiple effects (as an array of something) rather than two ways, one of which only supports a single effect for each type. 

Personally, I think that the FSA like approach is how I'd prefer to see it as it is familiar to redux users, so patterns like action creators (or effect creators, in this case) would be similar and familiar as well. 

Taking a step back, if you look at a structure like the one you linked to: https://github.com/matthieu-beteille/redux-data-fx/blob/b9e3b14f72445dcb48f4b646ff7f7cb059efafde/test/redux-data-fx.test.ts#L110-L136 each effect is an object with a single key whose key is important and whose value is important. This is not a familiar construct. Contrast that with:

```js
   fx({ value: 50 }, [ 
     { 
       effect: 'localStorage',
       set: { 
         a: 1, 
         b: 2 
       } 
     }, 
     { 
       effect: 'localStorage',
       set: { 
         c: 3, 
         d: 4 
       } 
     }, 
     { 
       effect: 'localStorage',
       set: { 
         e: 5, 
         f: 6 
       } 
     } 
   ]) 
```

The above is, in my opinion, more familiar, easier to (strongly) type and less complex. 

Note that you could use `type` as the key instead of `effect`, but that could get a little too close to an FSA. You could also take the FSA bit further and use a `payload` key for the data sent to the effect handler.

Any ways, your library looks great and it totally scratches an itch for me, so thank you for publishing it. Feel free to treat this issue as "please fix the types for batched effetcs" and/or close it out. Thanks!

**Comment by mt-bt** [comment]
That makes a lot of sense.

It definitely doesn't seem like a familiar construct when you want to run multiple effects. I didn't think that would be a common use case since you can implement your handler to accept a collection of items if needed.

That said, you're right the FSA like structure will definitely look much more familiar to redux users. I quite like the following format:

```javascript 
{ type: 'effectType', payload: { ... } } 
```

The only (small) downsides I see are:
- a little more verbose
- a little harder to test (it's very easy to check that a particular side effect is in the effect map)

I think I'll give it a go and implement this API change in the following weeks.

Thanks again for the kind words, and the ideas 😄

**Comment by mt-bt** [comment]
Solved by #2, closing
